# Supervised Learning - Building a Student Performace Prediction System


# Classification vs. Regression
The aim of this project is to predict how likely a student is to pass. Which type of supervised learning problem is this, classification or regression? Why?
Answer:
This project is a classification supervised learning problem because the variable to predict, i.e. if a student graduates or fails to graduate, is categorical. On this case this a dichotomous categorical variable where the only two possible values are "pass" or "fail".

### Overview:

1.Read the problem statement.

2.Get the dataset.

3.Explore the dataset.

4.Pre-processing of dataset.

5.Transform the dataset for building machine learning model.

6.Split data into train, test set.

7.Build Model.

8.Apply the model.

9.Evaluate the model.

10.Provide insights.

## Problem Statement 

Using Logistic Regression **predict the performance of student**. The classification goal is to predict whether the student will pass or fail.

## Dataset 

This data approach student achievement in secondary education of two Portuguese schools. The data attributes include student grades, demographic, social and school related features) and it was collected by using school reports and questionnaires. Two datasets are provided regarding the performance in Mathematics.

**Source:** https://archive.ics.uci.edu/ml/datasets/Student+Performance

# Question 1 - Exploring the Data (0.5 points)
*Read the dataset file using pandas. Take care about the delimiter.*

#### Answer:

In [185]:
import pandas as pd
import numpy as np

In [186]:
data=pd.read_csv('students-data.csv',';')
data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other ...   
3     GP   F   15       U     GT3       T     4     2   health  services ...   
4     GP   F   16       U     GT3       T     3     3    other     other ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

# Question 2 - drop missing values (0.5 points)
*Set the index name of the dataframe to **"number"**. Check sample of data to drop if any missing values are there.*
*Use .dropna() function to drop the NAs*

#### Answer:

In [187]:
#Set Index to number
data.index.name="number"

In [188]:
data.head(3)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  \
number                                                                         
0          GP   F   18       U     GT3       A     4     4  at_home  teacher   
1          GP   F   17       U     GT3       T     1     1  at_home    other   
2          GP   F   15       U     LE3       T     1     1  at_home    other   

       ... famrel freetime  goout  Dalc  Walc health absences G1 G2  G3  
number ...                                                               
0      ...      4        3      4     1     1      3        6  5  6   6  
1      ...      5        3      3     1     1      3        4  5  5   6  
2      ...      4        3      2     2     3      3       10  7  8  10  

[3 rows x 33 columns]

# Transform Data

## Question 3 (0.5 points)

*Print all the attribute names which are not numerical.*

**Hint:** check **select_dtypes()** and its **include** and **exclude** parameters.**

#### Answer:

In [189]:
data_non_num=data.select_dtypes(exclude='int64')
data_non_num.columns

Index(['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob',
       'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities',
       'nursery', 'higher', 'internet', 'romantic'],
      dtype='object')

In [190]:
data_non_num.dtypes

school        object
sex           object
address       object
famsize       object
Pstatus       object
Mjob          object
Fjob          object
reason        object
guardian      object
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
dtype: object

# Question 4 - Drop variables with less variance (0.5 points)

*Find the variance of each numerical independent variable and drop whose variance is less than 1. Use .var function to check the variance*

In [191]:
less_var=[]
for col in data.select_dtypes(exclude='object').columns:
    if data[col].var()<1:
        less_var.append(col)

In [192]:
less_var

['traveltime', 'studytime', 'failures', 'famrel', 'freetime', 'Dalc']

In [193]:
data.drop(less_var,inplace=True,axis=1)

In [194]:
data.head(3)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  \
number                                                                         
0          GP   F   18       U     GT3       A     4     4  at_home  teacher   
1          GP   F   17       U     GT3       T     1     1  at_home    other   
2          GP   F   15       U     LE3       T     1     1  at_home    other   

       ... higher internet romantic goout Walc health absences G1 G2  G3  
number ...                                                                
0      ...    yes       no       no     4    1      3        6  5  6   6  
1      ...    yes      yes       no     3    1      3        4  5  5   6  
2      ...    yes      yes       no     2    3      3       10  7  8  10  

[3 rows x 27 columns]

#### Variables with less variance are almost same for all the records. Hence, they do not contribute much for classification.

# Question 6 - Encode all categorical variables to numerical (0.5 points)

Take the list of categorical attributes(from the above result) and convert them into neumerical variables. After that, print the head of dataframe and check the values.

**Hint:** check **sklearn LabelEncoder()**

#### Answer:

In [195]:
from sklearn import preprocessing

In [196]:
le=preprocessing.LabelEncoder()

In [197]:
data[data_non_num.columns]=(data_non_num).apply(le.fit_transform)

In [198]:
data.head(3)

school  sex  age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  \
number                                                                        
0            0    0   18        1        0        0     4     4     0     4   
1            0    0   17        1        0        1     1     1     0     2   
2            0    0   15        1        1        1     1     1     0     2   

       ...  higher  internet  romantic  goout  Walc  health  absences  G1  G2  \
number ...                                                                      
0      ...       1         0         0      4     1       3         6   5   6   
1      ...       1         1         0      3     1       3         4   5   5   
2      ...       1         1         0      2     3       3        10   7   8   

        G3  
number      
0        6  
1        6  
2       10  

[3 rows x 27 columns]

# Question 7 - Convert the continuous values of grades into classes (1 point)

*Consider the values in G1, G2 and G3 with >= 10 as pass(1) and < 10 as fail(0) and encode them into binary values. Print head of dataframe to check the values.*

#### Answer:

In [199]:
for ind in range(0,data.shape[0]):
    if ((data.loc[ind]['G1'])>=10):
        data.loc[ind]['G1']=1
    else: 
        data.loc[ind]['G1']=0
    if ((data.loc[ind]['G2'])>=10):
        data.loc[ind]['G2']=1
    else: 
        data.loc[ind]['G2']=0
    if ((data.loc[ind]['G3'])>=10):
        data.loc[ind]['G3']=1
    else: 
        data.loc[ind]['G3']=0

In [201]:
data['G3'].value_counts()

1    265
0    130
Name: G3, dtype: int64

In [202]:
data.head(3)

school  sex  age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  \
number                                                                        
0            0    0   18        1        0        0     4     4     0     4   
1            0    0   17        1        0        1     1     1     0     2   
2            0    0   15        1        1        1     1     1     0     2   

       ...  higher  internet  romantic  goout  Walc  health  absences  G1  G2  \
number ...                                                                      
0      ...       1         0         0      4     1       3         6   0   0   
1      ...       1         1         0      3     1       3         4   0   0   
2      ...       1         1         0      2     3       3        10   0   0   

        G3  
number      
0        0  
1        0  
2        1  

[3 rows x 27 columns]

# Question 8 (0.5 points)

*Consider G3 is the target attribute and remaining all attributes as features to predict G3. Now, separate feature and target attributes into separate dataframes with X and y variable names.*

In [203]:
feature=data.drop('G3',axis=1)
feature.head(3)

school  sex  age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  \
number                                                                        
0            0    0   18        1        0        0     4     4     0     4   
1            0    0   17        1        0        1     1     1     0     2   
2            0    0   15        1        1        1     1     1     0     2   

       ...  nursery  higher  internet  romantic  goout  Walc  health  \
number ...                                                             
0      ...        1       1         0         0      4     1       3   
1      ...        0       1         1         0      3     1       3   
2      ...        1       1         1         0      2     3       3   

        absences  G1  G2  
number                    
0              6   0   0  
1              4   0   0  
2             10   0   0  

[3 rows x 26 columns]

In [204]:
target=data['G3']

# Question 9 - Training and testing data split (0.5 points)

# *So far, you have converted all categorical features into numeric values. Now, split the data into training and test sets with training size of 300 records. Print the number of train and test records.*

**Hint:** check **train_test_split()** from **sklearn**

#### Answer:

In [205]:
from sklearn.model_selection import train_test_split

In [206]:
X_train,X_test,Y_train,Y_test=train_test_split(feature,target,train_size=300,random_state=0)

C:\Users\Shreyaspatil\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2026: FutureWarning: From version 0.21, test_size will always complement train_size unless both are specified.
  FutureWarning)


In [207]:
print("Number of train records={0}\nNumber of test records={1}".format(X_train.shape[0],X_test.shape[0]))

Number of train records=300
Number of test records=95


# Question 10 - Model Implementation and Testing the Accuracy (0.5 points)

*Build a **LogisticRegression** classifier using **fit()** functions in sklearn. 
* You need to import both Logistic regression and accuracy score from sklearn*
#### Answer:

In [211]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [209]:
model = LogisticRegression()
model.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [233]:
Y_predict = model.predict(X_test)
model_score = model.score(X_test, Y_test)
print("Accuracy score of model for test data is",model_score)

Accuracy score of model for test data is 0.9263157894736842


# Question 11 - Print the intercept of the Logistic regression model (0.5 points)

The value of the intercepts are stored in the model itself. You can use .intercept_ function to do the same

In [218]:
print("Value of intercept for the model is",model.intercept_)

Value of intercept for the model is [0.4189925]


# Question 12 - Print the coefficients of the model (0.5 points) and name the coefficient which has the highest impact on the dependent variable (0.5 points)

Hint: Use .coef_ to get the coefficients and use pd.Dataframe to store the coefficients in a dataframe with column names same as the independent variable dataframe

In [219]:
model.coef_

array([[-0.22904771, -0.32495407, -0.15349906,  0.03377947, -0.09884194,
        -0.1800173 ,  0.03449812, -0.24024074,  0.01961358,  0.3495995 ,
         0.06360806,  0.22197777, -0.44019782, -0.21921146,  0.26552036,
        -0.18538049, -0.01038718,  0.77827836, -0.07820053, -0.43571062,
        -0.31118737,  0.52825515, -0.0677828 , -0.04138519,  1.15959276,
         3.99501176]])

In [226]:
coeff=pd.DataFrame(model.coef_,columns=feature.columns)
coeff

school       sex       age   address   famsize   Pstatus      Medu  \
0 -0.229048 -0.324954 -0.153499  0.033779 -0.098842 -0.180017  0.034498   

       Fedu      Mjob      Fjob    ...      nursery    higher  internet  \
0 -0.240241  0.019614  0.349599    ...    -0.010387  0.778278 -0.078201   

   romantic     goout      Walc    health  absences        G1        G2  
0 -0.435711 -0.311187  0.528255 -0.067783 -0.041385  1.159593  3.995012  

[1 rows x 26 columns]

<b> G2 has highest impact with coefficient = 3.99501176

# Question 13 - Predict the dependent variable for both training and test dataset (0.5 points)

Accuracy score() should help you to print the accuracies

In [234]:
Y_predict = model.predict(X_test)
model_score = model.score(X_test, Y_test)
print("Accuracy score of model for test data is",model_score)

Accuracy score of model for test data is 0.9263157894736842


In [236]:
Y_predict_train = model.predict(X_train)
model_score_train = model.score(X_train, Y_train)
print("Accuracy score of model for train data is",model_score_train)

Accuracy score of model for train data is 0.92
